### ISSUE TITLE
Android Health Connect API

### ISSUE THREAD

**Issue by 00pi** [home-assistant/android#2537]
**Is your feature request related to a problem? Please describe.**
For years I tried to sync my Xiaomi Smart Scale 2 to my Polar Flow account. Interfacing it with HA though ESPhome works as a charm, but there is no official write enabled API provided by Polar. Neither do I like to use any of the available fitness data syncing services such as SyncMyTracks as they request my login credentials and are subject to blockage from the providers at any time. Nor have I found any suitable service to directly populate my data and sync it to Polar Flow

**Describe the solution you'd like**
Check whether the Health Connect API can be implemented in the Android Companion App (or to avoid potential conflicts with Google: provide a separate Home Assistant Health Companion App). This would provide an interface to and from any fitness app that is going to join the Helath Connect API. And due to Googles power on the market, I expect them to join sooner or later.

**Describe alternatives you've considered, if any**
native Integrations for health related cloud services are not widely available, are subject to API limitations or blocking (https://github.com/home-assistant/core/issues/53833) were based on web scraping and thus removed from HA (https://github.com/home-assistant/architecture/issues/252)

**Additional context**
quote from (https://developer.android.com/guide/health-and-fitness/health-connect)
"Health Connect is an Android API and platform. It unifies data from multiple devices and apps into an ecosystem. For Android developers, it provides a single interface for reading and writing a user’s health and fitness data. For Android users, it offers a place for control over which apps have read and/or write access to different types of data. Health Connect also provides on-device storage.
[...]
Health Connect meets that need by unifying access to health and fitness data and empowering users to have control over their data. It provides users with a single way to access an overview of the health and fitness data they’re storing and sharing, along with access to privacy controls. Health Connect is an API that enables all our devices and apps across our ecosystem to work together."

**Comment by dshokouhi** [comment]
I don't see any current apps that support this API yet, lets try to find some apps that will supply the API with data so we can read from them. I know we can write data and then read it but we are not a fitness app so it wouldn't make sense for us to write the data. If we find any compatible apps lets comment on here so we can investigate this closer. From the apps perspective we will probably do sensor tracking to start.

**Comment by jpelgrom** [comment]
Would Home Assistant even be eligible to connect? While it might be possible to describe the functionality as something that is compliant, due to HA's nature the user can do anything with the data which is explicitly not allowed: https://developer.android.com/guide/health-and-fitness/health-connect-guidelines/health-connect-policy-requirement-faqs

> What are the approved use cases for Health Connect permissions?
For applications requesting access to any Health Connect permissions, approved use cases include fitness and wellness, rewards, fitness coaching, corporate wellness, medical care, health research, and games. Applications granted access to this permission **may not extend its use to undisclosed or non-permitted purposes**.

**Comment by dshokouhi** [comment]
We could probably try to make an argument for personal health research :)

**Comment by dshokouhi** [comment]
ok looks like first app on my phone to support this is Sleep as Android, no surprised there lol.

**Comment by acesyde** [comment]
Hi there,

Currently these apps support the Android Health Connect API

- Samsung Health
- Withings Health Mate
- Google fit

**Comment by jonathan-morton** [comment]
The beta came out with support for many more apps as well
https://android-developers.googleblog.com/2022/11/leading-health-and-fitness-apps-roll-out-health-connect-integrations.html

**Comment by dshokouhi** [comment]
right now the main question for us is, do we qualify for this API. Not every app is eligible

https://developer.android.com/guide/health-and-fitness/health-connect-guidelines/health-connect-policy-requirement-faqs

I do not know if we qualify for an approved use case here 🤔 

Also seems like we will need a security assessment as well.

**Comment by siikanen** [comment]
Any updates on this one? One be very interested importing my health connect days to HA

**Comment by dshokouhi** [comment]
> Any updates on this one?

no updates, until we know we qualify for the API we will probably not move forward with development.

With that said we do have Health Services sensors which allow you to pull the same data directly off a wearable. They are part of the same health API and also provide data to it.

**Comment by ndm13** [comment]
Just connected Fitbit since that's now supported, and noticed that their integration is write only. Maybe a read only HA integration would be more likely to be accepted?

**Comment by dshokouhi** [comment]
the security and permission requirements are the same, honestly its tough to tell before we attempt to do approval

**Comment by siikanen** [comment]
Is there any ongoing attempt to get the approval or are you planning to? If there's anything I can help with, let me know

**Comment by dshokouhi** [comment]
> Is there any ongoing attempt to get the approval or are you planning to?

we are all volunteers and work on whatever it is we want, you are more than welcome to read the docs and see what needs to be done.

**Comment by pacjo** [comment]
How about using [metriport](https://github.com/metriport/metriport) as an alternative?

**Comment by dshokouhi** [comment]
> How about using [metriport](https://github.com/metriport/metriport) as an alternative?

that looks like something that should be integrated into Home Assistant core as opposed to the app directly.

**Comment by pacjo** [comment]
You're probably right. I will check if issue like this is already open and if not open one.

**Comment by dshokouhi** [comment]
> You're probably right. I will check if issue like this is already open and if not open one.

you should not open an issue with HA core, instead a feature request on the forums.

https://community.home-assistant.io/c/feature-requests/13

**Comment by Hedda** [comment]
FYI, Google’s official Android Developers Blog has a new headlining post about Withings migrating to use Health Connect:

https://android-developers.googleblog.com/2023/03/withings-reduces-data-sync-code-with--health-and-fitness-api-health-connect.html

”***Withings reduces 50% of its data sync code by streamlining health and fitness API integrations with Health Connect***”

**Comment by PhillipRt** [comment]
All my fitness related apps are now connected to Health connect. Samsung Health, Google fit, Fitbit, Whoop. Implementing a read snyc would be great for a lot of my automations

**Comment by GOinfo-Ltd** [comment]
OnePlus oHealth syncs to Health Connect as well, and with Google Fit deprecating its API, this is becoming a must have.

**Comment by dshokouhi** [comment]
Initial set of sensors have been added and will be part of 2024.10.2-full once approved by Google in the [beta](https://play.google.com/apps/testing/io.homeassistant.companion.android). I will keep this issue open for a couple of weeks until Hacktoberfest is done in case anyone wants to try to add some more.

**Comment by dshokouhi** [comment]
Update to my recent comment: Unfortunately we were rejected by Google and due to an upcoming release have decided to temporarily disable the sensors so we can try again after the next release.  I'll update in here next time we try again.

![image](https://github.com/user-attachments/assets/a8f4d7bf-485d-4c56-96d0-71680b3a1a59)

**Comment by Webbeh** [comment]
Quite ironic that THEY refuse access for THAT reason...

**Comment by siikanen** [comment]
Oh my... I was already getting my hopes up for this after years of waiting but of course there's a final boss fight at the end

**Comment by dshokouhi** [comment]
> but of course there's a final boss fight at the end

😂 😂 

Submitted for approval again 🤞

**Comment by mapsindustrial** [comment]
When will we have news? I hope they are good!

**Comment by dshokouhi** [comment]
Good news looks like our beta build was approved! Please [join the beta](https://play.google.com/apps/testing/io.homeassistant.companion.android) if you have not already and update to 2024.11.1-full

Note: the change log was not updated yet but it will be next week :)

As Hacktoberfest is now completed I will close this as completed. In terms of adding additional sensors I am not sure its wise to add all the remaining ones in one final batch. I would say though maybe hold off for a couple weeks before making requests for sensors as I am sure others (myself included) will want to add some sensors for data we already have in Health Connect.

**Comment by mapsindustrial** [comment]
How can we see these sensors in Home Assistant?
It's a small tutorial possible?

Excuse my ignorance, thank you very much !

**Comment by dshokouhi** [comment]
> How can we see these sensors in Home Assistant?

1. First update to app to 2024.11.1-full as mentioned above
2. Enable like you do for any other app sensor. Open the app then: Settings > companion app > manage sensors > health connect and enable and grant permissions.

Also you need to be on Android 14 or higher for now

https://companion.home-assistant.io/docs/core/sensors#health-connect-sensors

**Comment by Blitzaflitza** [comment]
So far, only the weight and calorie features are working. When will other options, like a blood glucose sensor, be added? Health Connect has the potential for more than just tracking weight and calories.

**Comment by mapsindustrial** [comment]
> > ¿Cómo podemos ver estos sensores en Home Assistant?
> 
> 1. Primera actualización de la aplicación a 2024.11.1-completa como se mencionó anteriormente
> 2. Habilite como lo hace para cualquier otro sensor de aplicación. Abra la aplicación entonces: Configuración > aplicación complementaria > administrar sensores > salud conectar y habilitar y otorgar permisos.
> 
> También debe estar en Android 14 o superior por ahora
> 
> https://companion.home-assistant.io/docs/core/sensors#health-connect-sensors

It's a shame I'm still on Android 12 and for that reason I can't see it. 

Is there any news on when it will be available for older versions of Android? 

I am very interested in this project!

**Comment by dshokouhi** [comment]
> When will other options, like a blood glucose sensor, be added?

as mentioned in my comment an hour ago, were working on those. My own plan is to add sensors for which I already have data in health connect for. Blood Glucose is not one of those sensors in my own list but maybe others have it? 

I just submitted a small PR to add a heart rate sensor (#4784) so we can also confirm teh approval flow as these sensors are not like our others they require some manual effort to get things going.

We wont be able to add all sensors overnight but we can add them in chunks. Some patience will be needed but all are welcome to contribute  the sensors



> Is there any news on when it will be available for older versions of Android?

no updates on that yet but again anyone can contribute the change for that, the original author for HC did consider this but given at the time we did not know we would be approved they opted not to. https://github.com/home-assistant/android/pull/4452#discussion_r1640713278

**Comment by sirtwist** [comment]
Can we get READ_BLOOD_GLUCOSE added? I can test as my continuous glucose monitor writes to Health Connect.

**Comment by dshokouhi** [comment]
> Can we get READ_BLOOD_GLUCOSE added? I can test as my continuous glucose monitor writes to Health Connect.

I was planning on adding this one in the next round but since it seems to be pretty popular I have added it to my current PR #4785

**Comment by mapsindustrial** [comment]
Thank you for all the work behind this project. I would very much appreciate future compatibility with Android 11 or similar and adding sensors such as steps and sleep.

**Comment by dshokouhi** [comment]
> I would very much appreciate future compatibility with Android 11 or similar

Thats next on my list. Submitted a draft PR as it requires more testing (Wear OS and lower versions of Android 9-13). If you can please help by testing the debug APK from the link below and let me know if the health connect sensors are working. This refactor touched all sensors so quite a bit needs to be looked at again to make sure its all still good. As long as you see data populating and the app is not hanging/crashing that means the changes worked well. You must install the full version from the extracted archive, it will install side by side the production app with a distinguishable red icon. I recommend giving a unique name during onboarding so you can easily remove it later. In addition to health connect please check other sensors that you use. Make sure the device has play services and health connect is installed and all setup before proceeding.

https://github.com/home-assistant/android/actions/runs/11733612041

> and adding sensors such as steps and sleep.

This was already done as part of my most recent PR, see the PR for which additional sensors were added. The docs will be updated when Google approves the build

**Comment by mapsindustrial** [comment]
I just started the test a moment ago and everything seems to be going
great, so far everything is OK and I will let you know if there is any
news.

I am infinitely grateful for your work and great help.

This is exactly what I was looking for a long time and it gives great value
to connected health (and all the platforms that link to it, wear os, zepp
flow, etc.) and Home Assistant.
Un saludo.
Miguel Ángel Peinado Salas.

**Comment by dshokouhi** [comment]
> I just started the test a moment ago and everything seems to be going great, so far everything is OK and I will let you know if there is any news.

can you share a quick screenshot of the sensors with their data please? thats the only real confirmation we need to know its working :) previously the app would hang as soon as you opened it and/or enabled a sensor.

**Comment by mapsindustrial** [comment]
> > I just started the test a moment ago and everything seems to be going great, so far everything is OK and I will let you know if there is any news.
> 
> can you share a quick screenshot of the sensors with their data please? thats the only real confirmation we need to know its working :) previously the app would hang as soon as you opened it and/or enabled a sensor.

Of course, I attach the screenshot:

![Screenshot_20241108-015710](https://github.com/user-attachments/assets/2d6ee859-dedc-42d8-9765-0692343aa5a7)

Please feel free to ask me if I can help with anything else. Thank you so much.

**Comment by dshokouhi** [comment]
perfect! thanks for the screenshot and confirmation :)

**Comment by mapsindustrial** [comment]
thank you very much for everything!
Un saludo.
Miguel Ángel Peinado Salas.

**Comment by elpeterson** [comment]
> > > I just started the test a moment ago and everything seems to be going great, so far everything is OK and I will let you know if there is any news.
> > 
> > 
> > can you share a quick screenshot of the sensors with their data please? thats the only real confirmation we need to know its working :) previously the app would hang as soon as you opened it and/or enabled a sensor.
> 
> Of course, I attach the screenshot:
> 
> ![Screenshot_20241108-015710](https://private-user-images.githubusercontent.com/64039176/384187702-2d6ee859-dedc-42d8-9765-0692343aa5a7.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwMzk0NTcsIm5iZiI6MTczMTAzOTE1NywicGF0aCI6Ii82NDAzOTE3Ni8zODQxODc3MDItMmQ2ZWU4NTktZGVkYy00MmQ4LTk3NjUtMDY5MjM0M2FhNWE3LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA4VDA0MTIzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyYTgyOTMyMzNlNGZlMmJiMGZjYWVlNDgwNzk3YzlkYTQxNzI2YzM5NzlkMjE0ODJiMGMwNmU5ZTIxNWUwYWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.C6vFsRGIHnIRzP2JeoXj2Pls7YkgVY7iC-7tWMlZMDs)
> 
> Please feel free to ask me if I can help with anything else. Thank you so much.

What devices/apps are you using that is allowing you to pull in all those sensors from Health Connect? The only sensors I have are Active calories burned, Heart rate, Total calories burned, and Weight. But in Google Fit I pull in Sleep and Oxygen Saturation from a Sleep app, and then exercise data from Peloton

**Comment by dshokouhi** [comment]
> The only sensors I have are Active calories burned, Heart rate, Total calories burned, and Weight.

those are the new sensors mentioned in the PR merged earlier today (#4785), they will be part of the next beta update to the app. The user was testing a new PR to test functionality on lower android versions and the new sensors were there as a result.

**Comment by elpeterson** [comment]
> > The only sensors I have are Active calories burned, Heart rate, Total calories burned, and Weight.
> 
> those are the new sensors mentioned in the PR merged earlier today (#4785), they will be part of the next beta update to the app. The user was testing a new PR to test functionality on lower android versions and the new sensors were there as a result.

Ohh, I misread and saw an update in Home Assistant and thought I was missing something. Sorry about the confusion!

**Comment by sirtwist** [comment]
![Screenshot_20241107-234144](https://github.com/user-attachments/assets/515663b1-c015-425c-a54c-bcfbf8ec9fd3)
Thanks for adding blood glucose. My dashboard is patiently waiting for it. 😁

**Comment by jonathan-morton** [comment]
I'm running Android 13

> > I would very much appreciate future compatibility with Android 11 or similar
> 
> Thats next on my list. Submitted a draft PR as it requires more testing (Wear OS and lower versions of Android 9-13). If you can please help by testing the debug APK from the link below and let me know if the health connect sensors are working. This refactor touched all sensors so quite a bit needs to be looked at again to make sure its all still good. As long as you see data populating and the app is not hanging/crashing that means the changes worked well. You must install the full version from the extracted archive, it will install side by side the production app with a distinguishable red icon. I recommend giving a unique name during onboarding so you can easily remove it later. In addition to health connect please check other sensors that you use. Make sure the device has play services and health connect is installed and all setup before proceeding.
> 
> https://github.com/home-assistant/android/actions/runs/11733612041
> 
> > and adding sensors such as steps and sleep.
> 
> This was already done as part of my most recent PR, see the PR for which additional sensors were added. The docs will be updated when Google approves the build

It is working on my Android 13 phone, but the error message saying it is missing the permission needs some work. It just pops me back to the home screen. I added the permissions on the app installation screen and only realized I also had to make sure I had the health connect app installed (I did) and then also specifically in there give Home Assistant the permissions as well. Maybe some guided dialogs

**Comment by dshokouhi** [comment]
> It is working on my Android 13 phone, but the error message saying it is missing the permission needs some work. It just pops me back to the home screen. I added the permissions on the app installation screen and only realized I also had to make sure I had the health connect app installed (I did) and then also specifically in there give Home Assistant the permissions as well. Maybe some guided dialogs

actually once you have health connect installed and configured if you enable a sensor in app it will take you to the permission dialog we have. I have not hit the situation you are in, sounds like an edge case as health connect was not setup properly but its also tough to follow your description.

Just tested myself and as soon as I enable a sensor I am prompted to grant permission, the dialog prompt may not show up multiple times if you deny it initially to prevent apps from abusing permissions.

Either way that dialog has existed before these sensors were added so we can look into that in a separate issue. This issue is getting out of hand at this point so we may opt to close and ask users to open additional bugs/requests once this last PR is merged.

Appreciate the testing.

**Comment by dshokouhi** [comment]
The beta has been approved! For those who can enjoy the new sensors! Thanks to those who tested the PR to bring the functionality to lower versions of Android. My own extended testing has been going well there.

**Comment by mapsindustrial** [comment]
> The beta has been approved! For those who can enjoy the new sensors! Thanks to those who tested the PR to bring the functionality to lower versions of Android. My own extended testing has been going well there.

Great News !!! Thanks you very much !!!

**Comment by dshokouhi** [comment]
> It is working on my Android 13 phone, but the error message saying it is missing the permission needs some work. It just pops me back to the home screen. I added the permissions on the app installation screen and only realized I also had to make sure I had the health connect app installed (I did) and then also specifically in there give Home Assistant the permissions as well. Maybe some guided dialogs

After further testing on the PR this situation was indeed the case on lower devices. Was able to hunt down an old device running android 12 and reproduced the issue. Also fixed both issues regarding the permission prompt and snackbar. (note: if you deny any of these permissions twice the prompt will stop appearing as Android protects users from apps calling the dialog multiple times) Thanks again for testing :)

**Comment by wwelles** [comment]
@dshokouhi this is very exciting. Can you estimate when these changes may hit the prod build?

**Comment by dshokouhi** [comment]
> Can you estimate when these changes may hit the prod build?

Not sure on a date but after this next PR (#4790) we will push a new beta to get some final feedback before pushing to prod. We like to let the beta soak for a week before we make a prod build to give testers time to update and report back.

**Comment by wwelles** [comment]
Thanks @dshokouhi !

**Comment by Rigear** [comment]
First of all, amazing work! I'm loving the ability to be able to see this data and react to it, especially the blood glucose values. 

A question: Would it be possible to add a toggle to determine if it pulls from the [getMillimolesPerLiter](https://developer.android.com/reference/androidx/health/connect/client/units/BloodGlucose#getMillimolesPerLiter()) method vs the [getMilligramsPerDeciliter](https://developer.android.com/reference/androidx/health/connect/client/units/BloodGlucose#getMilligramsPerDeciliter()) one? 

I know I can do the math afterward the fact via a template sensor to convert it, it would just be nice to use my region's local units directly. 

Please let me know if you'd like me to open a new issue instead!

**Comment by dshokouhi** [comment]
> Please let me know if you'd like me to open a new issue instead!

that needs to be a separate request this issue will be locked after #4790 is merged

**Comment by dshokouhi** [comment]
> > Please let me know if you'd like me to open a new issue instead!
> 
> that needs to be a separate request this issue will be locked after #4790 is merged

@Rigear looks like the next version of HA core will add support for this natively: https://github.com/home-assistant/core/pull/129340 for the app we may just opt to support this when its out

**Comment by dshokouhi** [comment]
Going to lock this issue now. We now have a good baseline of sensors. Support was added for all versions of android for health connect, this will be available in the next beta. Work is in progress to add the other sensors. Please keep in mind it will take some time to add them, test them and get them approved. After the next beta soaks for some time we will cut a release to include them in production.

**Comment by dshokouhi** [comment]
Just an update, after receiving several back to back approvals we got rejected again. Google had asked us to remove everything. We also need to push out another app release. So we are going to disable Health Connect while we go for another production release. Once that release is out we will try again. If we get rejected again we will need to appeal which will take time to get a real person to get back to us. Stay tuned for more updates.

**Comment by dshokouhi** [comment]
Another update after numerous rejections and appeals we are not confident we will get access to Health Connect. Every rejection and/or appeal either lists or all almost all permissions. There are no guarantees even with a small amount of permissions it will be accepted. Even after our 3 previous approvals. We have yet to speak to a real person who can give us a straight answer as to the reasons for the rejection so we can correct the declaration or make code changes. The rejection reason has not changed either from up above. We will keep the code commented out for now if we can think of another approval method but at this point its not looking good.

**Comment by dshokouhi** [comment]
Just a quick update. We have tried again and got beta approval for a small set of recently approved permissions from our last appeal. We are now trying to go for production release and hopefully won't encounter any issues 🤞  . After that we should re-evaluate the other sensors and see if we can get approved for those or not too. Either way will provide another update here for the production release. You will want to be on 2025.1.2-full for the recently approved beta version. Hopefully we won't have to revert.....again.
